File: backend/app/graph/executor.py
"""图执行器 — 拓扑排序执行、并行节点、条件路由。

对标 LangGraph graph.invoke() + graph.astream()。
"""

from __future__ import annotations

import asyncio
import logging
import time
from typing import Any, AsyncGenerator

from app.graph.graph import GraphNode, StateGraph

logger = logging.getLogger("minicc.graph.executor")


class NodeResult:
    """单个节点的执行结果。"""
    def __init__(self, node_id: str, status: str = "completed", output: Any = None, error: str | None = None):
        self.node_id = node_id
        self.status = status
        self.output = output
        self.error = error
        self.duration_ms = 0.0


class GraphExecutor:
    """图执行器。

    用法:
        executor = GraphExecutor()
        result = await executor.invoke(graph, {"input": "hello"})
    """

    def __init__(self, tool_registry=None) -> None:
        self._registry = tool_registry
        self._cancel_event = asyncio.Event()
        self.results: dict[str, NodeResult] = {}

    def cancel(self) -> None:
        self._cancel_event.set()

    async def invoke(self, graph: StateGraph, state: dict[str, Any] | None = None) -> dict[str, Any]:
        """同步执行整个图，返回最终状态。"""
        async for _ in self.astream(graph, state):
            pass
        return self._aggregate_state(state or {})

    async def astream(self, graph: StateGraph, state: dict[str, Any] | None = None) -> AsyncGenerator[dict, None]:
        """流式执行，每个节点完成时 yield 事件。"""
        state = state or {}
        completed: set[str] = set()
        running: set[str] = set()
        node_map = {n.id: n for n in graph.nodes}
        adj: dict[str, list[str]] = {n.id: [] for n in graph.nodes}
        for edge in graph.edges:
            adj[edge.source_id].append(edge.target_id)

        # 拓扑排序后的节点列表
        from collections import deque
        in_deg = {n.id: 0 for n in graph.nodes}
        for edge in graph.edges:
            in_deg[edge.target_id] = in_deg.get(edge.target_id, 0) + 1

        ready = deque([nid for nid, d in in_deg.items() if d == 0])

        while ready and not self._cancel_event.is_set():
            # 获取当前批次所有可执行的节点
            batch = list(ready)
            ready.clear()

            # 并行执行
            tasks = []
            for nid in batch:
                tasks.append(self._execute_node(node_map[nid], state))
                running.add(nid)

            finished = await asyncio.gather(*tasks, return_exceptions=True)

            for nid, fut in zip(batch, finished):
                running.discard(nid)
                if isinstance(fut, Exception):
                    self.results[nid] = NodeResult(nid, status="error", error=str(fut))
                    yield {"type": "node_error", "node_id": nid, "error": str(fut)}
                else:
                    result: NodeResult = fut
                    self.results[nid] = result
                    state[nid] = result.output
                    yield {"type": "node_completed", "node_id": nid, "output": result.output}
                    completed.add(nid)

                # 更新下一个批次
                for target in adj.get(nid, []):
                    in_deg[target] -= 1
                    if in_deg[target] == 0:
                        ready.append(target)

        if self._cancel_event.is_set():
            yield {"type": "cancelled"}

    async def _execute_node(self, node: GraphNode, state: dict) -> NodeResult:
        """执行单个节点。"""
        start = time.monotonic()
        result = NodeResult(node_id=node.id)

        try:
            # 模拟执行 — G6 将替换为真实执行器
            await asyncio.sleep(0.1)
            output = f"[{node.node_type}] {node.label} executed"
            result.output = output
            result.status = "completed"
        except Exception as exc:
            result.status = "error"
            result.error = str(exc)

        result.duration_ms = (time.monotonic() - start) * 1000
        return result

    def _aggregate_state(self, state: dict) -> dict:
        """聚合所有节点结果到最终状态。"""
        final = dict(state)
        for nid, result in self.results.items():
            final[nid] = result.output if result.status == "completed" else result.error
        return final

File: internal/graph/executor.go
package graph

import (
	"context"
	"fmt"
	"log/slog"
	"strings"
	"sync"
	"time"

	"github.com/athenavi/minicc/internal/tools"
)

type NodeResult struct {
	NodeID     string      `json:"node_id"`
	Status     string      `json:"status"` // completed / error / skipped
	Output     string      `json:"output,omitempty"`
	Error      string      `json:"error,omitempty"`
	DurationMs int64       `json:"duration_ms"`
}

type GraphEvent struct {
	Type     string      `json:"type"` // node_started / node_completed / node_error / done
	NodeID   string      `json:"node_id,omitempty"`
	Output   string      `json:"output,omitempty"`
	Error    string      `json:"error,omitempty"`
	Progress float64     `json:"progress,omitempty"`
}

type Executor struct {
	mu      sync.RWMutex
	results map[string]*NodeResult
	events  []GraphEvent
}

func NewExecutor() *Executor {
	return &Executor{
		results: make(map[string]*NodeResult),
	}
}

func (ex *Executor) Execute(ctx context.Context, graph *StateGraph, initialState map[string]interface{}) (map[string]interface{}, []GraphEvent) {
	state := make(map[string]interface{})
	for k, v := range initialState {
		state[k] = v
	}

	ex.mu.Lock()
	ex.results = make(map[string]*NodeResult)
	ex.events = nil
	ex.mu.Unlock()

	nodeMap := make(map[string]GraphNode)
	for _, n := range graph.Nodes {
		nodeMap[n.ID] = n
	}

	// Build adjacency
	adj := make(map[string][]string)
	inDeg := make(map[string]int)
	for _, n := range graph.Nodes {
		adj[n.ID] = nil
		inDeg[n.ID] = 0
	}
	for _, e := range graph.Edges {
		adj[e.SourceID] = append(adj[e.SourceID], e.TargetID)
		inDeg[e.TargetID]++
	}

	// Start with nodes that have no dependencies
	var ready []string
	for id, d := range inDeg {
		if d == 0 {
			ready = append(ready, id)
		}
	}

	completed := make(map[string]bool)

	for len(ready) > 0 {
		// Execute ready nodes (sequentially for simplicity)
		var nextReady []string

		for _, nodeID := range ready {
			if completed[nodeID] {
				continue
			}

			node := nodeMap[nodeID]

			ex.addEvent(GraphEvent{Type: "node_started", NodeID: nodeID})

			start := time.Now()
			result := ex.executeNode(node, state)
			result.DurationMs = time.Since(start).Milliseconds()

			ex.mu.Lock()
			ex.results[nodeID] = &result
			ex.mu.Unlock()

			if result.Status == "completed" {
				state[nodeID] = result.Output
				completed[nodeID] = true
				ex.addEvent(GraphEvent{
					Type:   "node_completed",
					NodeID: nodeID,
					Output: result.Output,
				})
			} else {
				ex.addEvent(GraphEvent{
					Type:   "node_error",
					NodeID: nodeID,
					Error:  result.Error,
				})
			}

			// Check conditions and update downstream readiness
			for _, targetID := range adj[nodeID] {
				if !completed[targetID] {
					// Check edge conditions
					shouldRun := true
					for _, e := range graph.Edges {
						if e.SourceID == nodeID && e.TargetID == targetID && e.Condition != "" {
							if !evaluateCondition(e.Condition, result.Output) {
								shouldRun = false
								break
							}
						}
					}
					if shouldRun {
						// Check all predecessors are done
						allDone := true
						for _, e2 := range graph.Edges {
							if e2.TargetID == targetID {
								if !completed[e2.SourceID] {
									allDone = false
									break
								}
							}
						}
						if allDone && !completed[targetID] {
							nextReady = append(nextReady, targetID)
						}
					}
				}
			}
		}

		ready = nextReady
	}

	ex.addEvent(GraphEvent{
		Type:     "done",
		Progress: 1.0,
	})

	return state, ex.getEvents()
}

func (ex *Executor) executeNode(node GraphNode, state map[string]interface{}) NodeResult {
	result := NodeResult{NodeID: node.ID}

	switch node.NodeType {
	case NodeInput:
		if v, ok := state[node.ID]; ok {
			result.Output = fmt.Sprintf("%v", v)
		} else if v, ok := state["input"]; ok {
			result.Output = fmt.Sprintf("%v", v)
		} else {
			result.Output = fmt.Sprintf("[input] %s", node.Label)
		}
		result.Status = "completed"

	case NodeLLM:
		prompt, _ := node.Config["prompt"].(string)
		result.Output = fmt.Sprintf("[llm] %s (simulated: %s)", node.Label, truncateStr(prompt, 50))
		result.Status = "completed"

	case NodeTool:
		toolName, _ := node.Config["tool"].(string)
		result.Output = fmt.Sprintf("[tool] %s executed", toolName)
		result.Status = "completed"

	case NodeCondition:
		result.Output = fmt.Sprintf("[condition] %s", node.Label)
		result.Status = "completed"

	case NodeOutput:
		result.Output = fmt.Sprintf("[output] %s", node.Label)
		result.Status = "completed"

	default:
		result.Status = "error"
		result.Error = fmt.Sprintf("unknown node type: %s", node.NodeType)
	}

	return result
}

func (ex *Executor) addEvent(event GraphEvent) {
	ex.mu.Lock()
	defer ex.mu.Unlock()
	ex.events = append(ex.events, event)
}

func (ex *Executor) getEvents() []GraphEvent {
	ex.mu.RLock()
	defer ex.mu.RUnlock()
	events := make([]GraphEvent, len(ex.events))
	copy(events, ex.events)
	return events
}

func (ex *Executor) Results() map[string]*NodeResult {
	ex.mu.RLock()
	defer ex.mu.RUnlock()
	results := make(map[string]*NodeResult)
	for k, v := range ex.results {
		results[k] = v
	}
	return results
}

func evaluateCondition(condition, output string) bool {
	cond := strings.ToLower(strings.TrimSpace(condition))
	out := strings.ToLower(output)

	switch {
	case cond == "true", cond == "success":
		return true
	case strings.HasPrefix(cond, "contains:"):
		return strings.Contains(out, strings.TrimPrefix(cond, "contains:"))
	case strings.HasPrefix(cond, "equals:"):
		return out == strings.TrimPrefix(cond, "equals:")
	default:
		return out != ""
	}
}

// ── Tools ─────────────────────────────────────────────────────────────────

type CreateTool struct{}

func NewCreateTool() *CreateTool { return &CreateTool{} }
func (t *CreateTool) Name() string       { return "graph_create" }
func (t *CreateTool) Description() string { return "Create a StateGraph workflow from node/edge definitions." }

func (t *CreateTool) Execute(ctx context.Context, input map[string]interface{}) (map[string]interface{}, error) {
	name, _ := input["name"].(string)
	if name == "" {
		return nil, fmt.Errorf("name is required")
	}

	builder := NewBuilder()

	if nodes, ok := input["nodes"].([]interface{}); ok {
		for _, n := range nodes {
			if node, ok := n.(map[string]interface{}); ok {
				id, _ := node["id"].(string)
				label, _ := node["label"].(string)
				nodeType, _ := node["node_type"].(string)
				config, _ := node["config"].(map[string]interface{})
				builder.AddNode(id, label, NodeType(nodeType), config)
			}
		}
	}

	if edges, ok := input["edges"].([]interface{}); ok {
		for _, e := range edges {
			if edge, ok := e.(map[string]interface{}); ok {
				src, _ := edge["source_id"].(string)
				tgt, _ := edge["target_id"].(string)
				cond, _ := edge["condition"].(string)
				label, _ := edge["label"].(string)
				builder.AddEdge(src, tgt, cond, label)
			}
		}
	}

	if entry, ok := input["entry_point"].(string); ok {
		builder.SetEntryPoint(entry)
	}

	graph, result := builder.Compile()
	if !result.Valid {
		return nil, fmt.Errorf("compile errors: %s", strings.Join(result.Errors, "; "))
	}

	return map[string]interface{}{
		"output":           fmt.Sprintf("Graph '%s' created (%d nodes, %d edges)\nTopo: %s", name, len(graph.Nodes), len(graph.Edges), FormatTopo(result.TopologicalOrder)),
		"node_count":       len(graph.Nodes),
		"edge_count":       len(graph.Edges),
		"topological_order": result.TopologicalOrder,
	}, nil
}

type RunTool struct{}

func NewRunTool() *RunTool { return &RunTool{} }
func (t *RunTool) Name() string       { return "graph_run" }
func (t *RunTool) Description() string { return "Create and execute a StateGraph workflow in one step." }

func (t *RunTool) Execute(ctx context.Context, input map[string]interface{}) (map[string]interface{}, error) {
	name, _ := input["name"].(string)
	if name == "" {
		return nil, fmt.Errorf("name is required")
	}

	builder := NewBuilder()
	if nodes, ok := input["nodes"].([]interface{}); ok {
		for _, n := range nodes {
			if node, ok := n.(map[string]interface{}); ok {
				id, _ := node["id"].(string)
				label, _ := node["label"].(string)
				nodeType, _ := node["node_type"].(string)
				config, _ := node["config"].(map[string]interface{})
				builder.AddNode(id, label, NodeType(nodeType), config)
			}
		}
	}
	if edges, ok := input["edges"].([]interface{}); ok {
		for _, e := range edges {
			if edge, ok := e.(map[string]interface{}); ok {
				src, _ := edge["source_id"].(string)
				tgt, _ := edge["target_id"].(string)
				cond, _ := edge["condition"].(string)
				label, _ := edge["label"].(string)
				builder.AddEdge(src, tgt, cond, label)
			}
		}
	}
	if entry, ok := input["entry_point"].(string); ok {
		builder.SetEntryPoint(entry)
	}

	graph, compileResult := builder.Compile()
	if !compileResult.Valid {
		return nil, fmt.Errorf("compile errors: %s", strings.Join(compileResult.Errors, "; "))
	}

	initialState, _ := input["initial_state"].(map[string]interface{})
	executor := NewExecutor()
	state, events := executor.Execute(ctx, graph, initialState)

	slog.Info("graph executed", "name", name, "nodes", len(graph.Nodes), "events", len(events))

	return map[string]interface{}{
		"output":  fmt.Sprintf("Graph '%s' executed (%d nodes)\nStatus: completed", name, len(graph.Nodes)),
		"state":   state,
		"events":  events,
		"results": executor.Results(),
	}, nil
}

func RegisterTools(tr *tools.ToolRegistry) {
	tr.Register(NewCreateTool())
	tr.Register(NewRunTool())
}

func truncateStr(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n] + "..."
}
